icepack/statistics.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (C) 2022 by Daniel Shapero <shapero@uw.edu>
+#
+# This file is part of icepack.
+#
+# icepack is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# The full text of the license can be found in the file LICENSE in the
+# icepack source directory or at <http://www.gnu.org/licenses/>.
+
+from copy import deepcopy
+import firedrake
+import pyadjoint, pyadjoint.optimization.rol_solver
+
+
+class StatisticsProblem:
+    def __init__(self, simulation, likelihood, prior, controls):
+        r"""Initialize a data assimilation problem
+
+        Parameters
+        ----------
+        simulation
+            A function that takes in the controls and returns the observables
+        likelihood
+            A function that takes in the observables returns a form that can be
+            assembled to produce the value of the model-data misfit, i.e. the
+            negative logarithm of the likelihood function
+        prior
+            A function that takes in the controls and returns a form that can
+            be assembled to produce the value of the regularization, i.e. the
+            negative log prior
```

Comment:
Probably a very dumb question, but are regularisations ever a function of the state as well as the control? Perhaps if one did that then one wouldn't get posterior consistency

Reply:
I can't really think of any. I did wonder if I should devote some effort to future-proofing this now, but at the same time I didn't want to complicate the API for a use case that no one has yet. What do you think?

Reply:
Leave it until someone complains